prime-router/src/main/kotlin/tokens/JtiCache.kt

```diff
@@ -0,0 +1,41 @@
+package gov.cdc.prime.router.tokens
+
+import org.apache.logging.log4j.kotlin.Logging
+import java.time.OffsetDateTime
+
+/**
+ * The JTI is a standard claim in a JWT.  It is nonce string value that uniquely identifies a JWT.
+ * The JtiCache is used to check whether this JTI (and hence, this JWT) has been encountered before.
+ * If it has, the auth request is rejected, to prevent replay attacks.
+ *
+ * Note:  DPC project implements very similar -
+ * See https://github.com/CMSgov/dpc-app/blob/master/dpc-api/src/main/java/gov/cms/dpc/api/auth/jwt/IJTICache.java
+ */
+
+abstract class JtiCache: Logging {
+    abstract fun cleanupCache()
+
+    abstract fun insertIntoCache(jti: String, expiresAt: OffsetDateTime)
+
+    abstract fun isPresentInCache(jti: String): Boolean
+
+    /**
+     * Rather than having a timer based cache cleanup, this lazily only cleans up stale cache entries when its called.
+     */
```

Comment:
- Does the cleanup add blocking time to the requester? i.e. If there are lots of cache items to cleanup, could the requester be stuck waiting for an extended period of time?
- Could a malicious actor DDOS ReportStream by filling up the cache and causing token requests to timeout?

Reply:
Good questions.   I'm pretty sure the cleanup can't accumulate very easily, but I'm going to create a ticket to do some refactoring on this.